### PR TITLE
Site Plugins:

### PR DESCRIPTION
**Pull request by jdrossl**
- Add new API for getting & installing plugins in sites

### CODE REVIEW

File: src/main/api/2/studio-api-2.yaml [hunk 1]
```diff
@@ -3096,6 +3096,84 @@ paths:
...
+                    $ref: '#/components/schemas/ApiResponse'
+                  plugins:
+                    type: array
+                    items:
+                      $ref: '#/components/schemas/PluginRecord'
+        '400':
+          $ref: '#/components/responses/BadRequest'
+        '401':
+          $ref: '#/components/responses/Unauthorized'
+        '500':
```
Comment (sumerjabri) [inline src/main/api/2/studio-api-2.yaml]: Add 403?

File: src/main/api/2/studio-api-2.yaml [hunk 2]
```diff
@@ -3096,6 +3096,84 @@ paths:
...
+              schema:
+                type: object
+                properties:
+                  response:
+                    $ref: '#/components/schemas/ApiResponse'
+        '400':
+          $ref: '#/components/responses/BadRequest'
+        '401':
+          $ref: '#/components/responses/Unauthorized'
+        '409':
```
Comment (sumerjabri) [inline src/main/api/2/studio-api-2.yaml]: Add 403?

File: src/main/api/2/studio-api-2.yaml [hunk 3]
```diff
@@ -3777,6 +3855,37 @@ components:
...
+          type: string
+          format: date-time
+          description: The date when the plugin was installed
+        files:
+          type: array
+          description: The list of installed files
+          items:
+            $ref: '#/components/schemas/FileRecord'
+
+    FileRecord:
```
Comment (sumerjabri) [inline src/main/api/2/studio-api-2.yaml:5674]: Let's discuss this part over a call.

File: src/main/java/org/craftercms/studio/model/rest/ApiResponse.java [hunk 4]
```diff
@@ -120,6 +120,9 @@
         "Marketplace server is unreachable", "Check the configuration to make sure the Marketplace URL is correct",
         StringUtils.EMPTY);
 
+    public static final ApiResponse PLUGIN_ALREADY_INSTALLED =
+            new ApiResponse(40002, "Plugin is already installed", StringUtils.EMPTY, StringUtils.EMPTY);
+
     public static final ApiResponse CMIS_UNREACHABLE =
             new ApiResponse(50000, "CMIS server is unreachable", "Advise the user that the CMIS server is not " +
```
Comment (sumerjabri) [inline src/main/java/org/craftercms/studio/model/rest/ApiResponse.java]: Was this added to the list of error codes in the OAS docs?

File: src/main/resources/crafter/studio/studio-config.yaml [hunk 5]
```diff
@@ -787,3 +787,24 @@ studio.serverless.delivery.deployer.target.template.params:
...
+studio.marketplace.plugin.installable:
+  - engine
+# The name of the sub-folder to copy plugin files
+studio.marketplace.plugin.folder: plugins
+# The mapping of folders to copy files from the plugin folder to the site
+studio.marketplace.plugin.mapping:
+  - templates: templates
+  - static-assets: static-assets
+  - scripts/components: scripts/components
+  - scripts/filters: scripts/filters
```
Comment (sumerjabri) [inline src/main/resources/crafter/studio/studio-config.yaml]: Let's discuss.

**Review by sumerjabri — CHANGES_REQUESTED** [review]

File: src/main/api/2/studio-api-2.yaml (hunk 1 above)
Comment (jdrossl) [reply src/main/api/2/studio-api-2.yaml]: Right now there are no permissions defined for these new operations, any authenticated user can execute them

File: src/main/java/org/craftercms/studio/impl/v2/service/marketplace/MarketplaceServiceImpl.java [hunk 6]
```diff
@@ -71,4 +75,17 @@ public void createSite(CreateSiteRequest request) throws RemoteRepositoryNotFoun
...
     }
 
+    @Override
+    @ValidateParams
+    public List<PluginRecord> getInstalledPlugins(@ValidateStringParam(name = "siteId") String siteId)
+            throws MarketplaceException {
+        return marketplaceServiceInternal.getInstalledPlugins(siteId);
+    }
+
+    //TODO: Add specific permissions?
```
Comment (sumerjabri) [inline src/main/java/org/craftercms/studio/impl/v2/service/marketplace/MarketplaceServiceImpl.java]: Permission is "site member"

**Review by rart — COMMENTED** [review]
Looks good.

The only feedback I’ve got is if we could, for the `/installed` endpoint, use the same response model as marketplace search ([MarketplacePlugin](https://app.swaggerhub.com/apis/craftercms/studio/4.0.0.1#/MarketplacePlugin), [GitMarketplacePlugin](https://app.swaggerhub.com/apis/craftercms/studio/4.0.0.1#/GitMarketplacePlugin)) which includes media, compatibility, and other plugin details. This would allow for a more visual and rich listing and details UI. The properties of the new model are of use too, so maybe combine the models, or nest them, or be able to relate them somehow (e.g. pluginRecord.pluginDescriptor?)

Question: will local site plugins – i.e. plug-ins that were developed and deployed with the site, without coming from the marketplace – also be indexed, returned and managed by these APIs? An additional property to flag whether a record came from marketplace or is “local” would be appropriate.

**Comment by jdrossl** [comment]
@rart
Local plugins developed as part of the site will not be returned by this API, only plugins installed from the marketplace.

About merging the response I don't think it's a good idea for now, but we can consider it later. The reason is that this API returns data from a file stored in the site, if there are 10 plugins installed we would need to query the MP for each one of the descriptors.

**Comment by rart** [comment]
Does the info of the MarketplacePlugin models come from the yaml file or it comes from some sort of database? Installed plugins have their yaml file, right? Could we extract directly from the yaml?

**Comment by jdrossl** [comment]
We are not storing the descriptors right now

**Review by sumerjabri — CHANGES_REQUESTED** [review]
Roles allowed to install plugins:
* Developers
* Admins

Roles allowed to list plugins:
* Everyone in the site

Switch 4.0.0.1 to 4.0.0.4 for API OAS, UM, and other places.

Remove sha512 from FileRecord in OAS

File: src/main/api/2/studio-api-2.yaml [hunk 7]
```diff
@@ -3945,6 +3945,88 @@ paths:
         '500':
           $ref: '#/components/responses/InternalServerError'
 
+  /marketplace/installed:
+    get:
+      tags:
+        - marketplace
+      summary: Get the list of plugins installed in the given site
+      operationId: getInstalledPlugins
```
Comment (sumerjabri) [inline src/main/api/2/studio-api-2.yaml:3954]: Please add required permission in the spec, like:

```
description: 'Required Permission: "XXYYZZ"'
```

**Review by sumerjabri — CHANGES_REQUESTED** [review]

**Review by sumerjabri — APPROVED** [review]

**Pull request merged**